### PR TITLE
Update to latest PromiseKit

### PR DESCRIPTION
**Pull request by jakubknejzlik**
Hello guys 👋

There's bunch of code changes, but the tests finally passed! Thank you for your review in advance. Change requests expected :)

tldr:
- PromiseKit is completely refactored out as extension
- methods now return `CodableStoreRequest` which is going to be used for cancellation 
- any idead how to write `CodableStoreRequest.send` method nicer are welcome 🙏

### CODE REVIEW

File: CodableStore/Classes/Environment.swift [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import Foundation
 import PromiseKit
 
+
```
Comment (VilemKurz) [inline CodableStore/Classes/Environment.swift]: space

**Review by tomassliz — CHANGES_REQUESTED** [review]
I've discovered two issues in this update:

1. UI API is called on a background thread: `UIApplication.setNetworkActivityIndicatorVisible`
2. Deserialization of objects doesn't set correctly `dateDecodingStrategy` & `keyDecodingStrategy`. That's because casting on line 58 & 59 in `Serialization.swift` always fails:
```swift
public func deserialize<T: Decodable>() throws -> T {
    if let data = self as? T {
        return data
    } else {
        let decoder = JSONDecoder()

        // THIS TYPE CASTING ALWAYS FAILS
        decoder.dateDecodingStrategy = (T.self as? CustomDateDecodable.Type)?.dateDecodingStrategy ?? .iso8601
        decoder.keyDecodingStrategy = (T.self as? CustomKeyDecodable.Type)?.keyDecodingStrategy ?? .useDefaultKeys

        return try decoder.decode(T.self, from: self)
    }
}
```
But tests don't catch this issue so the should be updated too.

**Comment by jakubpetrik** [comment]
@tomassliz I've added a test to see if custom strategy is used or not and it indeed is. Can you please give some minimal example in which it is failing for you? Also network indicator should be now fixed.

**Review by jakubpetrik — DISMISSED** [review]

**Comment by 3ducation1** [comment]
@jakubpetrik I will try to post some example during this weekend.

**Comment by tomassliz** [comment]
@jakubpetrik I will try to post some example during this weekend.

**Comment by tomassliz** [comment]
@jakubpetrik I'm a little bit stuck. I can't replicate errors in tests but I have project which is failing. I can send you the project source with more info on email. Are you ok with that?

**Review by radimhalfar — APPROVED** [review]

**Comment by jakubpetrik** [comment]
> @jakubpetrik I'm a little bit stuck. I can't replicate errors in tests but I have project which is failing. I can send you the project source with more info on email. Are you ok with that?

@tomassliz  sure. :)

**Comment by tomassliz** [comment]
> @tomassliz sure. :)

@jakubpetrik OK, can you send me your e-mail please? Just write me on iCloud.com email joined with my GitHub username.

**Comment by jakubpetrik** [comment]
Ok. I can now confirm that issue with getting custom strategies still exists. 
Issue is when `T` is `Optional<CustomDateDecodable>`. I will try to fix that.

**Review by tomassliz — APPROVED** [review]

**Comment by tomassliz** [comment]
@jakubpetrik @jakubknejzlik I verified changes and it's ok now 👍  Shall we merge this PR?

**Review by jakubpetrik — APPROVED** [review]

**Comment by jakubpetrik** [comment]
Let's merge it please. @jakubknejzlik

**Pull request merged**